opt_einsum/paths.py

```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+
+
+def _ssa_optimize(inputs, output, sizes):
+    """
+    This has an interface similar to :func:`cheap` but produces a path with
+    static single assignment ids rather than recycled linear ids.
+    SSA ids are cheaper to work with and easier to reason about.
+    """
+    # A dim common to all terms might as well be an output dim.
+    inputs = list(map(frozenset, inputs))
+    output = frozenset(output) | frozenset.intersection(*inputs)
+
+    # Deduplicate shapes by eagerly computing Hadamard products.
+    remaining = {}  # key -> ssa_id
+    ssa_ids = itertools.count(len(inputs))
+    ssa_path = []
+    for ssa_id, key in enumerate(inputs):
+        if key in remaining:
+            ssa_path.append((remaining[key], ssa_id))
+            remaining[key] = next(ssa_ids)
+        else:
+            remaining[key] = ssa_id
+
+    # Compute footprints of each tensor.
+    footprints = {key: helpers.compute_size_by_dict(key, sizes)
+                  for key in remaining}
+
+    # Keep track of possible contraction dims.
+    dim_to_keys = defaultdict(set)
+    for key in remaining:
+        for dim in key - output:
+            dim_to_keys[dim].add(key)
+
+    # Keep track of the number of tensors using each dim; when the dim is no longer
+    # used it can be contracted. Since we specialize to binary ops, we only care about
+    # ref counts of >=2 or >=3.
+    dim_ref_counts = {
+        count: set(dim for dim, keys in dim_to_keys.items() if len(keys) >= count) - output
+        for count in [2, 3]}
+
+    # Find initial candidate contractions.
+    queue = []
+    for dim, keys in dim_to_keys.items():
+        keys = list(keys)
+        for i, k1 in enumerate(keys):
+            k2s = keys[:i]
+            _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue)
+
+    # Greedily contract pairs of tensors.
+    while queue:
+        cost, k1, k2, k12 = heapq.heappop(queue)
+        if k1 not in remaining or k2 not in remaining:
+            continue  # candidate is obsolete
+
+        ssa_id1 = remaining.pop(k1)
+        ssa_id2 = remaining.pop(k2)
+        for dim in k1 - output:
+            dim_to_keys[dim].remove(k1)
+        for dim in k2 - output:
+            dim_to_keys[dim].remove(k2)
+        ssa_path.append((ssa_id2, ssa_id1))
+        if k12 in remaining:
+            ssa_path.append((remaining[k12], next(ssa_ids)))
+        else:
+            footprints[k12] = helpers.compute_size_by_dict(k12, sizes)
+            for dim in k12 - output:
+                dim_to_keys[dim].add(k12)
+        remaining[k12] = next(ssa_ids)
+        _update_ref_counts(dim_to_keys, dim_ref_counts, k1 | k2 - output)
+
+        # Find new candidate contractions.
+        k1 = k12
+        k2s = set(k2 for dim in k1 for k2 in dim_to_keys[dim] if k2 != k1)
+        _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue)
+
+    # Greedily compute pairwise outer products.
```

Comment:
@jcmgray Am I doing this right? I don't really understand the outer product phase, and I'm not sure what the difference is between this and the simpler
```py
# Perform a final outer-product contraction.
ssa_path.append(tuple(remaining.values()))
return ssa_path
```

Reply:
The difference is just efficiency, although I agree it's quite a niche scenario. Doing pairwise outer products is generally faster than performing the entire outer product simultaneously with einsum (and whats more permits non-einsum backends). There's a test for this: ``test_can_optimize_outer_products``. 

I think you should be able to use the exact same costing/logic as the inner products, the only difference being that we only invoke checking all the outer products if there >1 remaining tensors but no remaining inner products (this is an approximate heuristic - sometimes the optimal path involves performing outer products first).

Reply:
Great, I've added tests cases to `test_can_optimize_outer_products`, refined the objective function, and removed the extraneous final contraction (moved to top of function as an edge case).

> use the exact same costing/logic as the inner products

Note the one difference: this phase optimizes for small inputs `(x1, x2)` independently, rather than optimizing for small outputs `x12`. By using a separable objective function, this phase has linear complexity; whereas if we used the same logic as for contractions this phase would have quadratic complexity. I'm not sure how much this matters in practice.

Reply:
Nice. Yeah if you are only left with outer products there should only be a few of them since they produce an exponentially large tensor (the most extreme case for example would be about <~30 tensors each of size 2). Your approach looks good however!

Reply:
So I just realized that this is an important case, e.g. if we have many pairs of contractions all lumped into a single einsum, e.g. "a,a,b,b,c,c,d,d,e,e". This results in linearly-many outer products (i.e. multiplies). So it's at least a little important to avoid quadratic growth here.